mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -0,0 +1,530 @@
+    {
+        let mut trading_bot_cfg = simple_market_maker_bot_ctx.trading_bot_cfg.lock().await;
+        // todo: check if clear is adapted, if i understand its keep the memory allocated for later usage.
+        trading_bot_cfg.clear();
+    }
+    // todo: cancel all pending orders
+}
+
+// This function check if coin is enabled and if check_balance is set to true and the balance is non-zero it's returned for later usage
+// e.g i want to use 50% of my KMD balance instead of max, will be used later.
+async fn coin_find_and_checks(
+    ticker: String,
+    key_trade_pair: String,
+    check_balance: bool,
+    ctx: &MmArc,
+) -> Result<MmNumber, MmError<OrderProcessingError>> {
+    let coin = match lp_coinfind(ctx, ticker.as_str()).await {
+        Ok(None) => {
+            warn!("{} not enabled - skipping for {}", ticker, key_trade_pair);
+            return MmError::err(OrderProcessingError::AssetNotEnabled);
+        },
+        Err(err) => {
+            warn!(
+                "err with {} - reason: {} - skipping for {}",
+                ticker, err, key_trade_pair
+            );
+            return MmError::err(OrderProcessingError::InternalCoinFindError);
+        },
+        Ok(Some(t)) => t,
+    };
+
+    if check_balance {
+        let coin_balance = match coin.my_balance().compat().await {
+            Ok(coin_balance) => coin_balance,
+            Err(err) => {
+                warn!(
+                    "err with balance: {} - reason: {} - skipping for {}",
+                    ticker,
+                    err.to_string(),
+                    key_trade_pair
+                );
+                return MmError::err(OrderProcessingError::BalanceInternalError);
+            },
+        };
+        if coin_balance.spendable.is_zero() {
+            warn!("balance for: {} is zero - skipping for {}", ticker, key_trade_pair);
+            return MmError::err(OrderProcessingError::BalanceIsZero);
+        }
+        return Ok(MmNumber::from(coin_balance.spendable));
+    }
+    Ok(MmNumber::default())
+}
+
+async fn vwap_apply(_calculated_price: &mut MmNumber) {}
+
+async fn cancel_single_order(ctx: &MmArc, uuid: Uuid) {
+    info!("cancelling single order with uuid: {}", uuid);
+    let resp = match cancel_order(ctx.clone(), json!({"uuid": uuid.to_string()})).await {
+        Ok(resp) => resp,
+        Err(_) => {
+            warn!("Couldn't cancel the order with uuid: {}", uuid);
+            return;
+        },
+    };
+
+    if resp.status() == StatusCode::OK {
+        info!("Order with uuid: {} successfully cancelled", uuid);
+    }
+}
+
+async fn checks_order_prerequisites(
+    ctx: &MmArc,
+    rates: &RateInfos,
+    cfg: &SimpleCoinMarketMakerCfg,
+    key_trade_pair: String,
+    uuid: Option<Uuid>,
+) -> OrderProcessingResult {
+    let cancel_functor = async move || {
+        if let Some(uuid) = uuid {
+            cancel_single_order(ctx, uuid).await
```

Comment:
Order cancellation might be an unexpected side effect of the fn named `checks_order_prerequisites`. It's worth moving the cancellation logic to the code that uses this.

Reply:
It will also allow removing the `uuid` argument.